### PR TITLE
DAOS-3150: Development of test case and test API for daos_run_io_conf.

### PR DESCRIPTION
**Pull request by ravalsam**
Test-tags: pr,hw

Signed-off-by: Samir Raval <samir.raval@intel.com>

### CODE REVIEW

File: src/tests/ftest/util/daos_io_conf.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    :avocado: recursive
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize a IoConfTestBase object."""
+        super(IoConfTestBase, self).__init__(*args, **kwargs)
+        self.pool = TestPool(self.context, self.log)
+
+    def setUp(self):
```
Comment (saurabhtandan) [inline src/tests/ftest/util/daos_io_conf.py]: Why is this needed if there is no modification to setUp()?

File: src/tests/ftest/util/daos_io_conf.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  Any reproduction of computer software, computer software documentation, or
+  portions thereof marked with this legend must also reproduce the markings.
+"""
+import os
+
+from apricot import TestWithServers
+from command_utils import ExecutableCommand
+from test_utils import TestPool
+
+class IoConfGen(ExecutableCommand):
```
Comment (saurabhtandan) [inline src/tests/ftest/util/daos_io_conf.py]: If daos_io_conf is a bin command with options and as you are already inheriting ExecutableCommand class, if you don't mind lets take full advantage of it. I think that whole code can be simplified and executed a bit differently. You can have a look at dfuse_utils.py/mdtest_utils.py for example.

Most of your gen_io_conf() params can go inside init() method of your IoConfGen() class. I am not familiar with the daos_io_conf bin, but if those are the options of that bin then it makes sense to have it in the init()

File: src/tests/ftest/rebuild/io_conf_run.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            Utility will create mixed of single/array data set.It takes the
+            snapshot of the record to verify the data later. During the test
+            it will exclude/add the specific/all targets from specific rank.
+            This verify the rebuild operation in loop for different targets
+            and also verify the data content.
+
+        Use Cases:
+            Verify rebuild with data verification.
+
+        :avocado: tags=all,medium,pr,hw,rebuild,iorebuild
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/rebuild/io_conf_run.py]: Yes, we do. So that NVMe is involve as well. Basically it also does rebuilding targets and even in ranks in some cases so having NVMe is better to cover the SPDK code path.

File: src/tests/ftest/util/daos_io_conf.py (hunk 1 above)
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: I think it's preferred to have it match with super class TestWithServers setUp(). Do you prefer to put it in side init()? or somewhere else ?

File: src/tests/ftest/util/daos_io_conf.py (hunk 2 above)
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: Well the init() will be more useful if we intended to use single test case with Avocado !mux, which is the case right now but looking for the future we need to use the same utility to the extend where we need to generate multiple files for single tests so that was the reason I created IoConfTestBase for run and create.

File: src/tests/ftest/util/daos_io_conf.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+               dkey_num,
+               akey_num,
+               rec_size,
+               obj_class,
+               filename]
+
+        #Run daos_gen_io_conf
+        #daos_gen_io_conf -g <rank_size> -t <tgt_size>-o [obj_num]
+        #-d [dkey_num] -a [akey_num] -s [rec_size]-O obj_class <file_name>
+        gen_io_conf = IoConfGen(" ".join(cmd))
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: It's difficult to keep up with other PR specially when those are in review. but I do not see any major impact on current implementation unless we are getting rid of self.prefix

File: src/tests/ftest/util/daos_io_conf.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        gen_io_conf.run()
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: Changing the gen_io_conf variable name to io_conf_gen as function name also the same. so it's not the self function instead its instance of  IoConfGen()

File: src/tests/ftest/util/daos_io_conf.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+        Args:
+            command (str): string of the command to be executed.
+            path (str, optional): path to location of command binary file.
+                Defaults to "".
+        """
+        super(IoConfGen, self).__init__("", command, path)
+        self.verbose = True
+        self.env = env
+
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/daos_io_conf.py]: I think what both @saurabhtandan and I are trying to convey is the inheritance of `ExecutableCommand` by `IoConfGen` is not being fully utilized.  

There are really three main features of the `ExecutableCommand` class:
1) It produces a command string including optional arguments when those arguments are specified
2) It provides a simplified one-method-call approach for reading these optional arguments from the test yaml by defining the yaml parameter key names and a fixed path in which to search for these keys
3) It provides a `run()` method for executing the command as a foreground or background process

All of these three features combine to define an object that is portable for use by other developers without needing to know too much about the command - they can just create the object, call `get_params()`, and call `run()`.

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: How do I pass the different argument to run_gen_io_conf() from the code? not from the yaml file

[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/daos_io_conf.py]: So assuming in the `__init__()` there is the following (and it passes `command="daos_gen_io_conf"` in the `super.__init__()`):
```
self.rank_size = FormattedParameter("-g {}")
self.target_size = FormattedParameter("-t {}")
self.record_size = FormattedParameter("-s {}", 3)
```

Then you can do the following:
```
io_conf = IoConfGen("/path/to/exe")
print(str(io_conf))
io_conf.rank_size.value = 10
print(str(io_conf))
io_conf.target_size.value = 5
print(str(io_conf))
```

The output from the print will look like this.  Note that the first `print()` only includes a `-s` command line argument because it is the only FormattedParameter object defined with a default.
```
/path/to/exe/daos_gen_io_conf -s 3
/path/to/exe/daos_gen_io_conf -g 10 -s 3
/path/to/exe/daos_gen_io_conf -g 10 -s 3 -t 5
```

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: Updated the code to use FormattedParameter class.

File: src/tests/ftest/util/daos_io_conf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/daos_io_conf.py]: I'm ok with that.  It was just a suggestion as most of classes are named after the executable they are representing.

File: src/tests/ftest/util/daos_io_conf.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+class IoConfTestBase(TestWithServers):
+    """Base rebuild test class.
+
+    :avocado: recursive
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize a IoConfTestBase object."""
+        super(IoConfTestBase, self).__init__(*args, **kwargs)
```
Comment (phender) [inline src/tests/ftest/util/daos_io_conf.py]: It looks like this definition really isn't needed.  It would only be required if the `IoConfTestBase` class needs to perform some addition actions currently not performed by the `TestWithServers.__init__()`.  Are there any extra steps required in the `__init__()` that are not specified here?  

If there was a need to run `daos_gen_io_conf` in one method and then `daos_run_io_conf` in a separate class method, I could see needing a `self.io_conf = None` statement included here, in which case it may make sense to have a `def setup_io_conf(self)` method.  Of course this could all be implemented if needed in the future.

File: src/tests/ftest/util/daos_io_conf.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        super(IoConfGen, self).__init__("/run/run_gen_io_conf/*",
+                                        "daos_gen_io_conf")
+        self.verbose = True
+        self.env = env
+        self.ranks = FormattedParameter("-g {}")
+        self.targets = FormattedParameter("-t {}")
+        self.obj_num = FormattedParameter("-o {}")
+        self.akeys = FormattedParameter("-a {}")
+        self.dkeys = FormattedParameter("-d {}")
+        self.re_size = FormattedParameter("-s {}")
```
Comment (phender) [inline src/tests/ftest/util/daos_io_conf.py]: Something to consider:  Since this variable name becomes the key in the yaml file, I find it beneficial for the variable name to be descriptive enough that someone reading only the yaml file can generally understand what impact the setting has to the test.  In my opinion, I think something like `rec_size` or `record_size` would be easier to understand in this case.  Ultimately the user can cross reference the command help with the argument string assigned to this variable to determine what impact the values has, but a descriptive name could avoid that extra work.

File: src/tests/ftest/util/daos_io_conf.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        """Create a ExecutableCommand object.
+
+        Uses Avocado's utils.process module to run a command str provided.
+
+        Args:
+            command (str): string of the command to be executed.
+            path (str, optional): path to location of command binary file.
+                Defaults to ""
+        """
+        super(IoConfGen, self).__init__("/run/run_gen_io_conf/*",
```
Comment (phender) [inline src/tests/ftest/util/daos_io_conf.py]: If consistency is something we want, we could consider a rename to the namespace argument.  Most, if not all, of the classes that inherit `ExecutableCommand` define a yaml path (`namespace`) that matches the executable name (`command`), e.g. `/run/ior` for `ior`, `/run/dmg` for `dmg`, etc.

File: src/tests/ftest/util/daos_io_conf.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                                        "daos_gen_io_conf")
```
Comment (phender) [inline src/tests/ftest/util/daos_io_conf.py]: It looks like the path was left off of this line:
```suggestion
                                        "daos_gen_io_conf", path)
```

File: src/tests/ftest/util/daos_io_conf.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        """Define a TestPool object."""
+        self.pool = TestPool(self.context, self.log)
+        self.pool.get_params(self)
+
+    def execute_io_conf_run_test(self):
+        """
+        Execute the rebuild test steps.
+        """
+        self.setup_test_pool()
+        io_conf = IoConfGen()
```
Comment (phender) [inline src/tests/ftest/util/daos_io_conf.py]: Should we be explicitly passing in a path here? Without it are we not relying on `daos_[gen|run]_io_conf` to be in the PATH?

For example:
```
io_conf = IoConfGen(self.bin)
```
or (if #1008 hasn't landed yet):
```
io_conf = IoConfGen(os.path.join(self.prefix, "bin"))
```

File: src/tests/ftest/util/daos_io_conf.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/daos_io_conf.py]: My apologies for not being more clear.  All I was trying to convey with that part of my comment was that #1008 introduces the `self.bin` definition to any `TestWithoutServer/TestWithServer` based class.  When it is available the use of `self.bin` would be preferred over `os.path.join(self.prefix, "bin")` as that is its definition.

File: src/tests/ftest/rebuild/io_conf_run.yaml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    - boro-H
+timeout: 900
+server_config:
+  name: daos_server
+  servers:
+    targets: 8
+    bdev_class: nvme
+    bdev_list: ["0000:5e:00.0","0000:5f:00.0"]
+pool:
+  scm_size: 14G
```
Comment (saurabhtandan) [inline src/tests/ftest/rebuild/io_conf_run.yaml]: If you are enabling 'nvme' above you will need to specify 'nvme_size' at pool create in order to take advantage of nvme.

File: src/tests/ftest/rebuild/io_conf_run.yaml [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    - boro-D
+    - boro-E
+    - boro-F
+    - boro-G
+    - boro-H
+timeout: 900
+server_config:
+  name: daos_server
+  servers:
+    targets: 8
```
Comment (saurabhtandan) [inline src/tests/ftest/rebuild/io_conf_run.yaml]: Not needed. if re-submitting please remove. Default is 8 in base file.

File: src/tests/ftest/rebuild/io_conf_run.yaml (hunk 12 above)
Comment (ravalsam) [reply src/tests/ftest/rebuild/io_conf_run.yaml]: In this pool creation is not used based on Python class, it's similar how daos_tests uses.

File: src/tests/ftest/rebuild/io_conf_run.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            Utility will create mixed of single/array data set.It takes the
+            snapshot of the record to verify the data later. During the test
+            it will exclude/add the specific/all targets from specific rank.
+            This verify the rebuild operation in loop for different targets
+            and also verify the data content.
+
+        Use Cases:
+            Verify rebuild with data verification.
+
+        :avocado: tags=all,medium,hw,rebuild,iorebuild
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/rebuild/io_conf_run.py]: will add full_regression. We don't want this test for each PR.

File: src/tests/ftest/rebuild/io_conf_run.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/python
+"""
+  (C) Copyright 2019 Intel Corporation.
```
Comment (saurabhtandan) [inline src/tests/ftest/rebuild/io_conf_run.py]: Minor: Shall we move on to change this to 2020. Only if you need to submit the PR.

File: src/tests/ftest/util/daos_io_conf.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/python
+"""
+  (C) Copyright 2019 Intel Corporation.
```
Comment (saurabhtandan) [inline src/tests/ftest/util/daos_io_conf.py]: Same as above.

File: src/tests/ftest/util/daos_io_conf.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+class IoConfTestBase(TestWithServers):
+    """Base rebuild test class.
+
+    :avocado: recursive
+    """
+
+    def setup_test_pool(self):
+        """Define a TestPool object."""
+        self.pool = TestPool(self.context, self.log)
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/tests/ftest/util/daos_io_conf.py]: I believe this is complaining that the constructor doesn't define the member variable.   Setting it to None would suffice

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/daos_io_conf.py]: I don't see this on latest commit
